File: app/src/main/java/pl/trikimusic/controller/core/bluetooth/WakeAdvertisementGate.kt
package pl.trikimusic.controller.core.bluetooth

class WakeAdvertisementGate(
    private val requiredSilenceMillis: Long = DEFAULT_REQUIRED_SILENCE_MILLIS,
) {
    private var lastAdvertisementMillis: Long? = null
    private var armed = false

    init {
        require(requiredSilenceMillis > 0L)
    }

    @Synchronized
    fun reset(nowMillis: Long) {
        lastAdvertisementMillis = nowMillis.takeIf { it >= 0L }
        armed = false
    }

    @Synchronized
    fun observeAdvertisement(nowMillis: Long): Boolean {
        if (nowMillis < 0L) {
            reset(nowMillis)
            return false
        }
        val previous = lastAdvertisementMillis
        if (previous == null || nowMillis < previous) {
            lastAdvertisementMillis = nowMillis
            armed = false
            return false
        }
        val mayConnect = armed || nowMillis - previous >= requiredSilenceMillis
        lastAdvertisementMillis = nowMillis
        return mayConnect
    }

    @Synchronized
    fun tryArm(nowMillis: Long): Boolean {
        val previous = lastAdvertisementMillis ?: return false
        if (nowMillis < previous) {
            reset(nowMillis)
            return false
        }
        if (armed || nowMillis - previous < requiredSilenceMillis) return false
        armed = true
        return true
    }

    @get:Synchronized
    val isArmed: Boolean
        get() = armed

    companion object {
        const val DEFAULT_REQUIRED_SILENCE_MILLIS = 5_000L
    }
}

File: windows/TrikiMusicController.Windows/Core/WakeAdvertisementGate.cs
namespace TrikiMusicController_Windows.Core;

public sealed class WakeAdvertisementGate
{
    public const long DefaultRequiredSilenceNanos = 5_000_000_000;

    private readonly object _sync = new();
    private readonly long _requiredSilenceNanos;
    private long? _lastAdvertisementNanos;
    private bool _armed;

    public WakeAdvertisementGate(long requiredSilenceNanos = DefaultRequiredSilenceNanos)
    {
        if (requiredSilenceNanos <= 0) throw new ArgumentOutOfRangeException(nameof(requiredSilenceNanos));
        _requiredSilenceNanos = requiredSilenceNanos;
    }

    public bool IsArmed
    {
        get
        {
            lock (_sync) return _armed;
        }
    }

    public void Reset(long nowNanos)
    {
        lock (_sync)
        {
            _lastAdvertisementNanos = nowNanos >= 0 ? nowNanos : null;
            _armed = false;
        }
    }

    public bool ObserveAdvertisement(long nowNanos)
    {
        lock (_sync)
        {
            if (nowNanos < 0 || _lastAdvertisementNanos is not long previous || nowNanos < previous)
            {
                _lastAdvertisementNanos = nowNanos >= 0 ? nowNanos : null;
                _armed = false;
                return false;
            }
            var mayConnect = _armed || nowNanos - previous >= _requiredSilenceNanos;
            _lastAdvertisementNanos = nowNanos;
            return mayConnect;
        }
    }

    public bool TryArm(long nowNanos)
    {
        lock (_sync)
        {
            if (_lastAdvertisementNanos is not long previous) return false;
            if (nowNanos < previous)
            {
                _lastAdvertisementNanos = nowNanos >= 0 ? nowNanos : null;
                _armed = false;
                return false;
            }
            if (_armed || nowNanos - previous < _requiredSilenceNanos) return false;
            _armed = true;
            return true;
        }
    }
}
